Dockerfile

```diff
@@ -20,6 +20,7 @@ RUN php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" \
     && php -r "unlink('composer-setup.php');"
 
 WORKDIR /lute
+VOLUME /lute/data
```

Comment:
I honestly don't know if a docker volume is better than a bind mount.  I have read that volumes are the preferred approach for Docker itself (ref https://docs.docker.com/storage/volumes/), but was wondering how that would affect users, especially those who want to get at their data, and how it would affect database backups.

Reply:
Notes from @disfated in Discord:

---

"So, there's no perfect solution. You have to choose between two evils."

* Evil #1 is volumes. Volumes are highly advised by docker, they can share data between containers, more importantly they provide [data population](https://docs.docker.com/storage/volumes/#populate-a-volume-using-a-container), so when a new user creates a container, /data directory is automatically set up. On the flip side, you can't access a volume on host machine - the only way is to spawn a temporary container with the volume attached to it. No fun...

* Meet evil #2. Bind mounts. These seem to be very handy as you can stick your host directory into the container. However they don't support data population, therefore you need to manually create initial /data directory structure for new users. Should be simple: we check if /data is empty, then copy files from some /data_init and the new user is good to go. The down side is that a user must explicitly specify the bind mount path, like -v ~/my_files/my_lute_data:/lute/data. If user doesn't do it right from the start, their data ends up inside the container, which is a terrible situation, because one doesn't ever persist sensitive data inside an (ephemeral) container. So, if you don't oblige user to specify host data directory this solution is even worse than volumes.

What I was thinking is that if you want Lute to become popular, you should target not only geeks, like us, but "regular people" as well. That way, there shouldn't be any need to physically deal with /data directory. Everything that users would require should be possible (in the future) to be done from within application's UI.

By the way, if I remember correctly, bind mounts take precedence over volumes. So, for your local setup, you are always able to stick your data directory from the host into the container with bind mount, i.e. volume should be ignored.

To conclude, if you ask me, I would go for volumes as the default with the ultimate goal of completely eliminating the need for end-users to directly interact with the filesystem. Devs can use bind mounts overriding volumes.

---

Reply:
The comments from @disfated are all great, but I still have some hesitations.  Before moving in this direction (ie using docker volumes) we’ll need to understand the following:

* how existing users will migrate to docker volumes.  Existing users have mounted volumes with their db, images, etc
* How backups are done. Currently the docker compose mounts an existing dir, eg /dropbox. Volumes aren’t good enough for this, IMO. I like the idea of db backups being zipped files on the host machine.  (But perhaps for docker deploys to cloud that idea doesn’t make sense anyway … I’ve always envisioned Lute as a desktop app on your personal machine, just like anki desktop.  Maybe there would need to be a separate “app” for cloud installs, but it might need more performance tests as some of the functions may spike the cpu on a shared machine.)
* The upgrade path - ie when a new docker image is created, how do users upgrade. This is probably trivial

Perhaps there should be two builds: one that uses volumes, and another that uses mounts ... I can't support the volumes option because I feel that non- or mid-techie users will rapidly get confused, and it will be overload for me.   Even for the current setup, with the docs written, I get people contacting me asking how to edit a .env file, or how to run console commands.

Lastly :-) I do want Lute to be as dead-easy to set up as possible, but I currently feel that that is best met by thinking more about Lute v3 written using Python.

Reply:
Thanks for the responses.

1. can continue to use bind mounts: super :-)

2. backups - I think my definition of "backup" hasn't been clear.  Currently, Lute has a job that exports the db as a timestamped backup, e.g `lute_backup_2023-09-14_001521.db.gz` to a specified folder.  I have it export to ~/Dropbox/LuteBackup on my machine.  The reason for this: even if my computer dies/is stolen (I travel a lot), I don't lose all of my Lute db and images, I can just reinstall Lute on another machine and copy the files over.  There are some tools (eg https://hub.docker.com/r/offen/docker-volume-backup) that do volume backups that might suffice, but you might know of another way.

3. Updates - yes, I wasn't referring to auto-updates, just the regular user updates as outlined at https://github.com/jzohrab/lute/wiki/Updating-Lute.  It's probably a non-issue, just `docker pull` the update and it should use the same volume.